src/main/scala-3/com/raquo/airstream/split/SplittableTypeMacros.scala

```diff
@@ -0,0 +1,185 @@
+  )(
+    using quotes: Quotes
+  ): Expr[MatchSplitObservable[Self, I, O1]] = {
+    import quotes.reflect.*
+    
+    matchSplitObservableExpr match {
+      case '{ MatchSplitObservable.build[Self, I, O]($observableExpr, $caseListExpr, $handlerMapExpr)} => innerHandleCaseImpl(observableExpr, caseListExpr, handlerMapExpr, casePfExpr, handleFnExpr)
+      case other => report.errorAndAbort("Macro expansion failed, please use `splitMatch` instead of creating new MatchSplitObservable explicitly")
+    }
+  }
+
+  private def innerHandleCaseImpl[Self[+_] <: Observable[_] : Type, I: Type, O: Type, O1 >: O : Type, A: Type, B: Type](
+    observableExpr: Expr[BaseObservable[Self, I]],
+    caseListExpr: Expr[List[PartialFunction[Any, Any]]],
+    handlerMapExpr: Expr[Map[Int, Function[Any, O]]],
+    casePfExpr: Expr[PartialFunction[A, B]],
+    handleFnExpr: Expr[Function[(B, Signal[B]), O1]]
+  )(
+    using quotes: Quotes
+  ): Expr[MatchSplitObservable[Self, I, O1]] = {
+    val caseExprList = exprOfListToListOfExpr(caseListExpr)
+
+    val nextCaseExprList = casePfExpr.asExprOf[PartialFunction[Any, Any]] :: caseExprList
+
+    val nextCaseListExpr = listOfExprToExprOfList(nextCaseExprList)
+
+    '{ MatchSplitObservable.build[Self, I, O1]($observableExpr, $nextCaseListExpr, ($handlerMapExpr + ($handlerMapExpr.size -> $handleFnExpr.asInstanceOf[Function[Any, O1]]))) }
+  }
+
+  private def exprOfListToListOfExpr(
+    pfListExpr: Expr[List[PartialFunction[Any, Any]]]
+  )(
+    using quotes: Quotes
+  ): List[Expr[PartialFunction[Any, Any]]] = {
+    import quotes.reflect.*
+    
+    pfListExpr match {
+      case '{ $headExpr :: (${tailExpr}: List[PartialFunction[Any, Any]]) } =>
+        headExpr :: exprOfListToListOfExpr(tailExpr)
+      case '{ Nil } => Nil
+      case _ => report.errorAndAbort("Macro expansion failed, please use `handleCase` instead of modify MatchSplitObservable explicitly")
+    }
+    
+  }
+
+  private def listOfExprToExprOfList(
+    pfExprList: List[Expr[PartialFunction[Any, Any]]]
+  )(
+    using quotes: Quotes
+  ): Expr[List[PartialFunction[Any, Any]]] = {
+    import quotes.reflect.*
+    
+    pfExprList match
+      case head :: tail => '{ $head :: ${listOfExprToExprOfList(tail)} }
+      case Nil => '{ Nil }
+  }
+
+  private inline def toSplittableOneObservable[Self[+_] <: Observable[_], O](
+    parentObservable: BaseObservable[Self, (Int, Any)],
+    handlerMap: Map[Int, Function[Any, O]]
+  ): Self[O] = {
+    parentObservable.matchStreamOrSignal(
+      ifStream = _.splitOne(_._1) { case (idx, (_, b), dataSignal) =>
+        val bSignal = dataSignal.map(_._2)
+        handlerMap.apply(idx).apply(b -> bSignal)
+      },
+      ifSignal = _.splitOne(_._1) { case (idx, (_, b), dataSignal) =>
+        val bSignal = dataSignal.map(_._2)
+        handlerMap.apply(idx).apply(b -> bSignal)
+      }
+    ).asInstanceOf[Self[O]] // #TODO[Integrity] Same as FlatMap/AsyncStatusObservable, how to type this properly?
```

Comment:
In your public API, you have `toStream` and `toSignal` methods, but you have a common implementation for them - `toSplittableOneObservable`, calling into `toSplittableOneObservable`. The last two methods require the Self type, which is where the type problem starts.

Airstream's underlying `splitOne` method is not a single method on Observable / Self, but rather we have two separate `splitOne` methods, one for streams, and one for signals. And so, they don't work with the `Self` type, requiring something like `matchStreamOrSignal` to use from a generalized Self method.

In principle, we could change Airstream to have only one unified `splitOne` method, implemented in a single SplitOneObservable value class with a Self type param, similar to e.g. BooleanObservable. However, the implementation of such a `splitOne` method would need to do something similar to `matchStreamOrSignal`, except with the output type `B <: Self`, to satisfy your method's desired output type. Perhaps add a new `matchStreamOrSignalSelf` method with such a constraint, as it may be useful in other cases as well?

There is still a minor problem with that – instead of doing `asInstanceOf`, you would need to throw an (effectively unreachable) exception like `matchStreamOrSignal` does, because the compiler thinks that we can't exhaustively match Observable as EventStream | Signal, because even though there are no other subtypes of Observable, I can not make Observable a sealed trait, because there is no way I can put the entire contents of EventStream and Signal types into one scala file for ergonomic reasons.

Still, I think the best option is to unify the `splitOne` method this way. Then you can call into a single splitOne method with the output Self type matching the input type – that will eliminate the need for `matchStreamOrSignal` in your macrro implementation. The next best option, if you don't want to mess with unifying `splitOne`, is to only implement `matchStreamOrSignalSelf` as I suggested, and just use that from your macros instead of `matchStreamOrSignal` – that should fix the types just as well. Lastly, as a last resort, you could also rewrite your macros to have separate implementations for signals and streams instead of one common implementation. But that actually sounds like more work, and would leave us with a bit of redundant code. NBD, but option  number 2 is probably superior in every way.

Reply:
This note was more of a todo/reminder rather than a question that needs answer right away. I'm fine with the way it is right now, but I can try to solve this if you think it is necessary.

Reply:
Eh, it's fine as-is, I'll try to get rid of this asInstanceOf if / when I unify the splitOne method.